### PR TITLE
Editorial: reword usage of rest arguments in built-in functions

### PR DESCRIPTION
**Pull request by devsnek**
- Specify that arguments covered by rest args aren't filled by `undefined`
- Remove steps which duplicate the definitions of rest arguments in algorithm bodies
- Remove most usage (ignoring the horror in Array and Date) of "(actual) number of (actual) arguments"

### CODE REVIEW

File: spec.html [hunk 1]
```diff
@@ -23106,7 +23106,7 @@ <h1>ECMAScript Standard Built-in Objects</h1>
...
+  <p>Unless otherwise specified in the description of a particular function, if a built-in function or constructor is given fewer arguments than the function is specified to require, the function or constructor shall behave exactly as if it had been given sufficient additional arguments, each such argument being the *undefined* value. Such missing arguments, if not covered by rest arguments, are considered to be &ldquo;not present&rdquo; and may be identified in that manner by specification algorithms. In the description of a particular function, the terms &ldquo;`this` value&rdquo; and &ldquo;NewTarget&rdquo; have the meanings given in <emu-xref href="#sec-built-in-function-objects"></emu-xref>.</p>
```
Comment (ljharb) [inline spec.html]: If you have a function with a signature of `(a, b, ...c)`:

`arguments.length` | a | b | c.length
--|--|--|--
0 | not present | not present | `0`
1 | present | not present | `0`
2 | present | present | `0`
3 | present | present | `1`

Given this table, it makes sense to me that `c` is always "present", even if it's empty by virtue of insufficient arguments being provided, so this change makes sense to me.

File: spec.html [hunk 2]
```diff
@@ -24033,8 +24033,7 @@ <h1>Object.assign ( _target_, ..._sources_ )</h1>
         <p>The `assign` function is used to copy the values of all of the enumerable own properties from one or more source objects to a _target_ object. When the `assign` function is called, the following steps are taken:</p>
         <emu-alg>
           1. Let _to_ be ? ToObject(_target_).
-          1. If only one argument was passed, return _to_.
-          1. Let _sources_ be the List of argument values starting with the second argument.
+          1. If _sources_ is empty, return _to_.
```
Comment (ljharb) [inline spec.html]: is `sources` a list, or an array (as it would be in code)? if the latter, is "is empty" defined for arrays?

File: spec.html [hunk 3]
```diff
@@ -24616,9 +24615,8 @@ <h1>Function.prototype.call ( _thisArg_, ..._args_ )</h1>
           1. Let _func_ be the *this* value.
           1. If IsCallable(_func_) is *false*, throw a *TypeError* exception.
           1. Let _argList_ be a new empty List.
```
Comment (ljharb) [inline spec.html]:
```suggestion
```

File: spec.html [hunk 4]
```diff
@@ -27890,16 +27891,11 @@ <h1>Properties of the String Constructor</h1>
...
           1. Let _elements_ be a new empty List.
-          1. Let _nextIndex_ be 0.
-          1. Repeat, while _nextIndex_ &lt; _length_
-            1. Let _next_ be _codeUnits_[_nextIndex_].
+          1. For each element _next_ of _codeUnits_, in ascending index order, do
             1. Let _nextCU_ be ? ToUint16(_next_).
             1. Append _nextCU_ to the end of _elements_.
-            1. Increase _nextIndex_ by 1.
-          1. Return the String value whose code units are, in order, the elements in the List _elements_. If _length_ is 0, the empty string is returned.
+          1. Return the String value whose code units are, in order, the elements in the List _elements_. If the number of elements in _codeUnits_ is 0, the empty string is returned.
```
Comment (ljharb) [inline spec.html]: can we use "If _codeUnits_ is empty, the empty string is returned", and maybe split this step up into an if/then two-step?

File: spec.html [hunk 5]
```diff
@@ -27908,18 +27904,13 @@ <h1>String.fromCharCode ( ..._codeUnits_ )</h1>
...
-          1. Repeat, while _nextIndex_ &lt; _length_
-            1. Let _next_ be _codePoints_[_nextIndex_].
+          1. For each element _next_ of _codePoints_, ascending index order, do
             1. Let _nextCP_ be ? ToNumber(_next_).
             1. If SameValue(_nextCP_, ! ToInteger(_nextCP_)) is *false*, throw a *RangeError* exception.
             1. If _nextCP_ &lt; 0 or _nextCP_ &gt; 0x10FFFF, throw a *RangeError* exception.
             1. Append the elements of the UTF16Encoding of _nextCP_ to the end of _elements_.
-            1. Increase _nextIndex_ by 1.
-          1. Return the String value whose code units are, in order, the elements in the List _elements_. If _length_ is 0, the empty string is returned.
+          1. Return the String value whose code units are, in order, the elements in the List _elements_. If the number of elements in _codePoints_ is 0, the empty string is returned.
```
Comment (ljharb) [inline spec.html]: ditto

File: spec.html (hunk 4 above)
Comment (ljharb) [reply spec.html]: I think it's useful to keep an explicit mention of the empty string case.

File: spec.html [hunk 6]
```diff
@@ -32778,7 +32762,7 @@ <h1>Runtime Semantics: IterableToList ( _items_, _method_ )</h1>
         <h1>%TypedArray%.of ( ..._items_ )</h1>
         <p>When the `of` method is called with any number of arguments, the following steps are taken:</p>
         <emu-alg>
-          1. Let _len_ be the actual number of arguments passed to this function.
+          1. Let _len_ be the number of elements in _items_.
           1. Let _items_ be the List of arguments passed to this function.
```
Comment (jmdyck) [inline spec.html]: You presumably want to delete the `Let _items_ be ...` step.

File: spec.html (hunk 4 above)
Comment (jmdyck) [reply spec.html]: You could make it a NOTE-step.

Comment (ljharb) [reply spec.html]: A note seems like it’d apply to the whole algorithm, instead of just this step.

File: spec.html [hunk 7]
```diff
@@ -27890,16 +27891,11 @@ <h1>Properties of the String Constructor</h1>
         <h1>String.fromCharCode ( ..._codeUnits_ )</h1>
         <p>The `String.fromCharCode` function may be called with any number of arguments which form the rest parameter _codeUnits_. The following steps are taken:</p>
         <emu-alg>
-          1. Let _codeUnits_ be a List containing the arguments passed to this function.
-          1. Let _length_ be the number of elements in _codeUnits_.
           1. Let _elements_ be a new empty List.
-          1. Let _nextIndex_ be 0.
-          1. Repeat, while _nextIndex_ &lt; _length_
-            1. Let _next_ be _codeUnits_[_nextIndex_].
+          1. For each element _next_ of _codeUnits_, in ascending index order, do
```
Comment (jmdyck) [inline spec.html]: Converting the 'while' loop to a 'for' loop (and similarly below) appears to be orthogonal to the main point of this PR. You might want to pull it out to a separate commit, if not a separate PR.

File: spec.html (hunk 1 above)
Comment (jmdyck) [inline spec.html]: I don't think it suffices to simply insert "if not covered by rest arguments", because we haven't defined 'rest argument' (or parameter) in the context of a built-in function's signature, or what it means to be "covered by" one.

Also,  I don't think you can delete specific steps like `Let _sources_ be the List of argument values starting with the second argument` without specifying the general mechanism instead.

So I think I'd like to see this paragraph expanded into a specification of the things that appear in built-ins' parameter lists (including square-brackets and ellipsis), and the way that parameters get their values for a given invocation (i.e., using values, if any, from the _argumentsList_ of the [[Call]]).  (This could possibly be a new clause under "Algorithm Conventions", or even something in [9.3.1](https://tc39.github.io/ecma262/#sec-built-in-function-objects-call-thisargument-argumentslist), but I think it might make more sense here.)

File: spec.html (hunk 4 above)
Comment (jmdyck) [reply spec.html]: A NOTE-*step* applies just to the point in the algorithm at which it occurs. (Scan the spec for `NOTE:`.)
